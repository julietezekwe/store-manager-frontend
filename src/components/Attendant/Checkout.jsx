/* eslint-disable no-restricted-globals */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getAllCartProducts, handleCheckout } from '../../actions/salesAction';

export class Checkout extends Component {
  state = { totalPrice: 0 };

  async componentDidMount() {
    if (JSON.parse(localStorage.getItem(`${this.props.login.user.id}`))) {
      const carts = JSON.parse(
        localStorage.getItem(`${this.props.login.user.id}`),
      );
      await this.props.getAllCartProducts(carts.cartNo);
    }
  }

  handleRemoveFromCart = (id) => {
    const carts = JSON.parse(
      localStorage.getItem(`${this.props.login.user.id}`),
    );

    for (let i = 0; i < carts.cartNo.length; i += 1) {
      if (Number(carts.cartNo[i]) === Number(id)) {
        carts.cartNo.splice(i, 1);
      }
    }
    const cartContent = JSON.stringify(carts);
    localStorage.setItem(`${this.props.login.user.id}`, cartContent);
    localStorage.setItem('cartCount', carts.cartNo.length);
    window.location.reload();
  };

  onQuantityChange = (quantity, id, price) => {
    if (quantity < 1 || isNaN(quantity)) quantity = 0;
    document.getElementById(`Q${id}`).innerHTML = quantity;
    const currentPrice = document.getElementById(`${id}`).innerHTML;
    const newPrice = quantity * price;

    document.getElementById(`${id}`).innerHTML = newPrice;
    let totalPrice = Number(document.getElementById('totalPrice').innerHTML);
    totalPrice = totalPrice - currentPrice + newPrice;
    this.setState({ totalPrice });
  };

  render() {
    const { cart } = this.props.sales;

    return (
      <div className="col-8">
        <div
          className="card col-12"
          style={{
            float: 'none',
            textAlign: 'center',
            marginLeft: 'auto',
            marginRight: 'auto',
            marginTop: '35px',
          }}
        >
          <div className="card col-9" id="alert" />
          <table className="table" style={{ width: '100%' }} id="checkout">
            <thead>
              <tr>
                <th>S/N</th>
                <th>Product Name</th>
                <th>Description</th>
                <th>ID</th>
                <th />
                <th>Quantity</th>
                <th>Amount (&#8358;)</th>
                <th>Action</th>
              </tr>
            </thead>
            <tbody>
              {cart
                && cart.cartProductDetails.map(product => (
                    <tr key={product[0].id}>
                      <td>{product[0].id}</td>
                      <td>{product[0].productname}</td>
                      <td>{product[0].description}</td>
                      <td>{product[0].id}</td>
                      <td>
                        <input
                          type="number"
                          min="1"
                          style={{ textAlign: 'center', width: '50px' }}
                          onChange={e => this.onQuantityChange(
                            e.target.value,
                            product[0].id,
                            product[0].price,
                          )
                          }
                        />{' '}
                      </td>
                      <td id={`Q${product[0].id}`}>1</td>
                      <td id={product[0].id}>{product[0].price}</td>
                      <td>
                        <i
                          className="far fa-trash-alt trigger"
                          onClick={() => this.handleRemoveFromCart(product[0].id)
                          }
                        />
                      </td>
                    </tr>
                ))}
            </tbody>
            <tfoot>
              <tr>
                <td colSpan="5" />
                <td>Total Price </td>
                <td id="totalPrice">
                  {this.state.totalPrice !== 0
                    ? this.state.totalPrice
                    : cart.totalPrice}
                </td>
              </tr>
              <tr>
                <td colSpan="2">
                  <button
                    className="button"
                    onClick={() => this.props.handleCheckout(this.props.login.user.id)
                    }
                  >
                    Checkout Now
                  </button>
                </td>
                <td colSpan="4" />
                <td colSpan="2" />
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  login: state.login,
  sales: state.sales,
});
export default connect(
  mapStateToProps,
  { getAllCartProducts, handleCheckout },
)(Checkout);
