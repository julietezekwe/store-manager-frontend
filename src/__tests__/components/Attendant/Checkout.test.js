/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable no-undef */
import { Checkout } from '../../../components/Attendant/Checkout.jsx';

sinon.stub(window.location, 'reload');
let wrapper;
const props = {
  getAllCartProducts: () => Promise.resolve(),
  handleCheckout: () => Promise.resolve(),
  sales: {
    cart: {
      cartProductDetails: [[{
        id: 1,
        productname: 'shoes',
        description: 'fine',
        quantity: 10,
        price: 1200,
        created_at: new Date(),
      }]],
    },
  },
  login: {
    user: {
      id: 3,
      role: 'attendant',
      email: 'chidimma@yahoo.com',
      name: 'chidimma',
      username: 'chidimma',
    },
  },
};
beforeEach(() => {
  wrapper = mount(<Checkout {...props} />);
});
describe('Checkout Component', () => {
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('calls handleRemoveFromCart()', () => {
    localStorage.setItem(3, '{"cartNo":[2, 6]}');
    sinon.spy(wrapper.instance(), 'handleRemoveFromCart');
    wrapper.instance().handleRemoveFromCart(1);
    expect(wrapper.instance().handleRemoveFromCart.calledOnce)
      .toEqual(true);
  });
  it('calls onQuantityChange()', () => {
    sinon.spy(wrapper.instance(), 'onQuantityChange');
    // wrapper.instance().onQuantityChange(1).debug();
    console.log(wrapper.instance().onQuantityChange(2, 2, 1000).debug());
    // expect(wrapper.instance().onQuantityChange.calledOnce)
    //   .toEqual(true);
    // expect(wrapper.instance().onFormFieldChange.calledWith(1, 2, 200));
  });
});
