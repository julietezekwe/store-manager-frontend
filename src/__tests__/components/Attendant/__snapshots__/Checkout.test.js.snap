// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout Component should match snapshot 1`] = `
<Checkout
  getAllCartProducts={[Function]}
  handleCheckout={[Function]}
  login={
    Object {
      "user": Object {
        "email": "chidimma@yahoo.com",
        "id": 3,
        "name": "chidimma",
        "role": "attendant",
        "username": "chidimma",
      },
    }
  }
  sales={
    Object {
      "cart": Object {
        "cartProductDetails": Array [
          Array [
            Object {
              "created_at": 2019-02-23T17:57:41.282Z,
              "description": "fine",
              "id": 1,
              "price": 1200,
              "productname": "shoes",
              "quantity": 10,
            },
          ],
        ],
      },
    }
  }
>
  <div
    className="col-8"
  >
    <div
      className="card col-12"
      style={
        Object {
          "float": "none",
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginTop": "35px",
          "textAlign": "center",
        }
      }
    >
      <div
        className="card col-9"
        id="alert"
      />
      <table
        className="table"
        id="checkout"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <thead>
          <tr>
            <th>
              S/N
            </th>
            <th>
              Product Name
            </th>
            <th>
              Description
            </th>
            <th>
              ID
            </th>
            <th />
            <th>
              Quantity
            </th>
            <th>
              Amount (â‚¦)
            </th>
            <th>
              Action
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            key="1"
          >
            <td>
              1
            </td>
            <td>
              shoes
            </td>
            <td>
              fine
            </td>
            <td>
              1
            </td>
            <td>
              <input
                min="1"
                onChange={[Function]}
                style={
                  Object {
                    "textAlign": "center",
                    "width": "50px",
                  }
                }
                type="number"
              />
               
            </td>
            <td
              id="Q1"
            >
              1
            </td>
            <td
              id={1}
            >
              1200
            </td>
            <td>
              <i
                className="far fa-trash-alt trigger"
                onClick={[Function]}
              />
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td
              colSpan="5"
            />
            <td>
              Total Price 
            </td>
            <td
              id="totalPrice"
            />
          </tr>
          <tr>
            <td
              colSpan="2"
            >
              <button
                className="button"
                onClick={[Function]}
              >
                Checkout Now
              </button>
            </td>
            <td
              colSpan="4"
            />
            <td
              colSpan="2"
            />
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</Checkout>
`;
